// oip-hdmw/lib/networks/litecoin.js
'use strict';

var _insightExplorer = require('insight-explorer');

var _coininfo = require('coininfo');

var _coininfo2 = _interopRequireDefault(_coininfo);

var _bip44Constants = require('bip44-constants');

var _bip44Constants2 = _interopRequireDefault(_bip44Constants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var litecoinFeePerKb = 100000;

var n = _coininfo2.default.litecoin.main.toBitcoinJS();

module.exports = {
	name: 'litecoin',
	displayName: 'Litecoin',
	ticker: 'LTC',
	satPerCoin: 1e8,
	feePerKb: litecoinFeePerKb,
	feePerByte: litecoinFeePerKb / 1024,
	maxFeePerByte: 100,
	minFee: litecoinFeePerKb,
	dust: 54600,

	txVersion: 1,

	explorer: new _insightExplorer.Insight('https://litesight.failover.alexandria.io/api'),

	getExtraBytes: function getExtraBytes(options) {
		return;
	},

	network: {
		bip32: {
			public: n.bip32.public,
			private: n.bip32.private
		},
		slip44: _bip44Constants2.default.LTC,
		messagePrefix: '\x18Litecoin Signed Message:\n',
		pubKeyHash: n.pubKeyHash,
		scriptHash: n.scriptHash,
		wif: n.wif
	}
};
